// chrome-devtools-mcp-for-extension/build/src/cli.js
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
export const cliOptions = {
    browserUrl: {
        type: 'string',
        description: 'Connect to a running Chrome instance using port forwarding. For more details see: https://developer.chrome.com/docs/devtools/remote-debugging/local-server.',
        alias: 'u',
        coerce: (url) => {
            new URL(url);
            return url;
        },
    },
    headless: {
        type: 'boolean',
        description: 'Whether to run in headless (no UI) mode.',
        default: false,
    },
    executablePath: {
        type: 'string',
        description: 'Path to custom Chrome executable.',
        conflicts: 'browserUrl',
        alias: 'e',
    },
    isolated: {
        type: 'boolean',
        description: 'If specified, creates a temporary user-data-dir that is automatically cleaned up after the browser is closed.',
        default: false,
    },
    customDevtools: {
        type: 'string',
        description: 'Path to custom DevTools.',
        hidden: true,
        conflicts: 'browserUrl',
        alias: 'd',
    },
    channel: {
        type: 'string',
        description: 'Specify a different Chrome channel that should be used. The default is the stable channel version.',
        choices: ['stable', 'canary', 'beta', 'dev'],
        conflicts: ['browserUrl', 'executablePath'],
    },
    loadExtension: {
        type: 'string',
        description: 'Load an unpacked Chrome extension from the specified directory path.',
        conflicts: 'browserUrl',
    },
    loadExtensionsDir: {
        type: 'string',
        description: 'Load all unpacked Chrome extensions from the specified directory. Each subdirectory with manifest.json will be loaded as an extension.',
        conflicts: 'browserUrl',
    },
    loadSystemExtensions: {
        type: 'boolean',
        description: 'Automatically discover and load extensions installed in the system Chrome profile. This includes extensions from Chrome Web Store and sideloaded extensions.',
        default: false,
        conflicts: 'browserUrl',
    },
    chromeProfile: {
        type: 'string',
        description: 'Specify Chrome profile name (e.g., "Default", "Profile 1", "Profile 2"). If not specified, uses last_used from Local State. Only effective when --loadSystemExtensions is true.',
        conflicts: 'browserUrl',
    },
    userDataDir: {
        type: 'string',
        description: 'Specify a custom user data directory for Chrome to use instead of the default. Auto-detected if not specified.',
        conflicts: 'browserUrl',
    },
    logFile: {
        type: 'string',
        describe: 'Path to a file to write debug logs to. Set the env variable `DEBUG` to `*` to enable verbose logs. Useful for submitting bug reports.',
    },
    projectRoot: {
        type: 'string',
        description: 'Explicitly specify the project root directory for profile isolation. Overrides MCP roots/list. Useful when roots/list is not available.',
        conflicts: 'browserUrl',
    },
};
export function parseArguments(version, argv = process.argv) {
    const yargsInstance = yargs(hideBin(argv))
        .scriptName('npx chrome-devtools-mcp@latest')
        .options(cliOptions)
        .check(args => {
        // Auto-configuration for zero-config setup
        if (!args.channel && !args.browserUrl && !args.executablePath) {
            args.channel = 'stable';
        }
        return true;
    })
        .example([
        ['$0', 'Zero-config startup: auto-detects profile and bookmarks'],
        [
            '$0 --browserUrl http://127.0.0.1:9222',
            'Connect to an existing browser instance',
        ],
        ['$0 --channel beta', 'Use Chrome Beta installed on this system'],
        ['$0 --channel canary', 'Use Chrome Canary installed on this system'],
        ['$0 --channel dev', 'Use Chrome Dev installed on this system'],
        ['$0 --channel stable', 'Use stable Chrome installed on this system'],
        [
            '$0 --loadSystemExtensions',
            'Auto-discover and load system Chrome extensions',
        ],
        [
            '$0 --loadExtensionsDir ./extensions --loadSystemExtensions',
            'Load both development and system extensions',
        ],
        ['$0 --logFile /tmp/log.txt', 'Save logs to a file'],
        ['$0 --help', 'Print CLI options'],
    ]);
    return yargsInstance
        .wrap(Math.min(120, yargsInstance.terminalWidth()))
        .help()
        .version(version)
        .parseSync();
}
