// chrome-devtools-mcp-for-extension/build/src/plugin-api.js
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
/**
 * Registry for managing MCP tools.
 * Allows dynamic registration and querying of tools.
 */
export class ToolRegistry {
    tools = new Map();
    categories = new Map();
    /**
     * Register a single tool.
     * @throws Error if a tool with the same name already exists
     */
    register(tool) {
        if (this.tools.has(tool.name)) {
            throw new Error(`Tool "${tool.name}" is already registered`);
        }
        this.tools.set(tool.name, tool);
        // Track by category
        const category = tool.annotations.category;
        if (!this.categories.has(category)) {
            this.categories.set(category, new Set());
        }
        this.categories.get(category).add(tool.name);
    }
    /**
     * Register multiple tools at once.
     */
    registerBatch(tools) {
        for (const tool of tools) {
            this.register(tool);
        }
    }
    /**
     * Get all registered tools.
     */
    getAll() {
        return Array.from(this.tools.values());
    }
    /**
     * Get tools by category.
     */
    getByCategory(category) {
        const names = this.categories.get(category);
        if (!names)
            return [];
        return Array.from(names)
            .map(name => this.tools.get(name))
            .filter(Boolean);
    }
    /**
     * Get a specific tool by name.
     */
    get(name) {
        return this.tools.get(name);
    }
    /**
     * Check if a tool is registered.
     */
    has(name) {
        return this.tools.has(name);
    }
    /**
     * Get count of registered tools.
     */
    get size() {
        return this.tools.size;
    }
    /**
     * Unregister a tool by name.
     * Returns true if the tool was removed, false if it didn't exist.
     */
    unregister(name) {
        const tool = this.tools.get(name);
        if (!tool)
            return false;
        this.tools.delete(name);
        const category = tool.annotations.category;
        this.categories.get(category)?.delete(name);
        return true;
    }
    /**
     * Clear all registered tools.
     */
    clear() {
        this.tools.clear();
        this.categories.clear();
    }
}
/**
 * Plugin loader for dynamically loading plugins.
 */
export class PluginLoader {
    plugins = new Map();
    registry;
    log;
    config;
    constructor(registry, log = console.error, config = {}) {
        this.registry = registry;
        this.log = log;
        this.config = config;
    }
    /**
     * Load a plugin from a module path or package name.
     * @param moduleId - Path to module or npm package name
     */
    async load(moduleId) {
        try {
            this.log(`[plugins] Loading plugin: ${moduleId}`);
            // Dynamic import
            const module = await import(moduleId);
            const plugin = module.default || module.plugin || module;
            if (!plugin.id || !plugin.register) {
                this.log(`[plugins] Invalid plugin (missing id or register): ${moduleId}`);
                return false;
            }
            if (this.plugins.has(plugin.id)) {
                this.log(`[plugins] Plugin already loaded: ${plugin.id}`);
                return false;
            }
            // Create plugin context
            const ctx = {
                registry: this.registry,
                log: msg => this.log(`[${plugin.id}] ${msg}`),
                config: this.config,
            };
            // Register the plugin
            await plugin.register(ctx);
            this.plugins.set(plugin.id, plugin);
            this.log(`[plugins] Loaded: ${plugin.name} v${plugin.version} (${plugin.id})`);
            return true;
        }
        catch (error) {
            this.log(`[plugins] Failed to load ${moduleId}: ${error instanceof Error ? error.message : String(error)}`);
            return false;
        }
    }
    /**
     * Load multiple plugins from environment variable or config.
     * @param pluginIds - Comma-separated list of plugin module IDs
     */
    async loadFromList(pluginIds) {
        const ids = pluginIds
            .split(',')
            .map(id => id.trim())
            .filter(Boolean);
        const loaded = [];
        const failed = [];
        for (const id of ids) {
            const success = await this.load(id);
            if (success) {
                loaded.push(id);
            }
            else {
                failed.push(id);
            }
        }
        return { loaded, failed };
    }
    /**
     * Unload a plugin by ID.
     */
    async unload(pluginId) {
        const plugin = this.plugins.get(pluginId);
        if (!plugin)
            return false;
        try {
            if (plugin.unload) {
                await plugin.unload();
            }
            this.plugins.delete(pluginId);
            this.log(`[plugins] Unloaded: ${pluginId}`);
            return true;
        }
        catch (error) {
            this.log(`[plugins] Failed to unload ${pluginId}: ${error instanceof Error ? error.message : String(error)}`);
            return false;
        }
    }
    /**
     * Get list of loaded plugins.
     */
    getLoaded() {
        return Array.from(this.plugins.values());
    }
}
